velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -114,37 +157,105 @@ void applyTypedWithInstance(
   auto rawOffsets = baseArray->rawOffsets();
   auto indices = arrayDecoded.indices();
 
-  rows.applyToSelected([&](auto row) {
-    auto size = rawSizes[indices[row]];
-    auto offset = rawOffsets[indices[row]];
-    auto search = searchDecoded.valueAt<T>(row);
-
-    auto instance = instanceDecoded.valueAt<int64_t>(row);
+  if (elementsDecoded.isIdentityMapping() && !elementsDecoded.mayHaveNulls() &&
+      searchDecoded.isConstantMapping() &&
+      instanceDecoded.isConstantMapping()) {
+    auto instance = instanceDecoded.valueAt<int64_t>(0);
     VELOX_USER_CHECK_NE(
         instance,
         0,
         "array_position cannot take a 0-valued instance argument.");
 
-    int startIndex = instance > 0 ? 0 : size - 1;
-    int endIndex = instance > 0 ? size : -1;
-    int step = instance > 0 ? 1 : -1;
-    instance = std::abs(instance);
-
-    int i;
-    for (i = startIndex; i != endIndex; i += step) {
-      if (!elementsDecoded.isNullAt(offset + i) &&
-          elementsDecoded.valueAt<T>(offset + i) == search) {
-        --instance;
-        if (instance == 0) {
-          flatResult.set(row, i + 1);
-          break;
+    if constexpr (std::is_same_v<bool, T>) {
+      auto rawElements = elementsDecoded.values<uint64_t>();
+
+      auto search = bits::isBitSet(searchDecoded.values<uint64_t>(), 0);
+
+      rows.applyToSelected([&](auto row) {
+        auto size = rawSizes[indices[row]];
+        auto offset = rawOffsets[indices[row]];
+
+        int startIndex = instance > 0 ? 0 : size - 1;
+        int endIndex = instance > 0 ? size : -1;
+        int step = instance > 0 ? 1 : -1;
+        instance = std::abs(instance);
+
+        int i;
+        for (i = startIndex; i != endIndex; i += step) {
+          if (bits::isBitSet(rawElements, offset + i) == search) {
+            --instance;
+            if (instance == 0) {
```

Comment:
can you reduce code repetition by moving the code to a templated function 
that is templated on the mode (FastBool, FastNotBool, SlowPath)
it looks like differences are at limited locations
- reading search and compare. 
```
if(conditions of fast path){
if(bool..){
RunCore<BoolFast>(..);
}else{
RunCore<FastNotBool>(..); 
}}else{
RunCore<SlowPath>
}
```
```
template< Mode T>
RunCore<T> {
       auto rawElements = elementsDecoded.values<uint64_t>();
      
      //  this is different based on mode
      T search =readConst<T>();

      rows.applyToSelected([&](auto row) {
        auto size = rawSizes[indices[row]];
        auto offset = rawOffsets[indices[row]];

        int startIndex = instance > 0 ? 0 : size - 1;
        int endIndex = instance > 0 ? size : -1;
        int step = instance > 0 ? 1 : -1;
        instance = std::abs(instance);
        for (auto i = startIndex; i != endIndex; i += step) {
                     //  this is different based on mode
                      if (check<bool>(search, offset, i,...) {
                      }
         }
         ...etc
```

Same for the other case when there is no instance.
I even then the case when there is no instance can be reduce to the case of instance =1.

Reply:
Thank you for this suggestion! I agree it would be great if we can reduce the code size through reusing. 

So I tried to combine the non-boolean and boolean fast paths into a function template, but had some difficulties. Specifically, the problem is that the type of rawElements are different between the two cases and the type is not exactly T (i.e., for the case where T is bool, the type of rawElements is const uint64_t*). 

I then tried to combine the regular (slow) path for primitive types and for complex types into a function template, but the size of this template is about the same as the sum of the two separate code pieces (because I need to add two if-constexpr branches and the function needs quite a few parameters.)

I also tried to make the no-instance case reuse the code for the with-instance case (i.e., make instance = 1), but the performance of the no-instance case decreased from over 80% to about 60% in the benchmark.